### PR TITLE
Several improvements and simplifications for hyperparameter optimization

### PR DESCRIPTION
**Pull request by mseeger**
*Issue #, if available:*

*Description of changes:*
A number of improvements and simplifications of hyperparameter optimization code:
  - Basic improvements in FIFOScheduler
  - Increase timeouts for ssh connections
  - SimpleProfiler
  - Cleanup GPFIFOSearcher, GPHyperbandSearcher
  - Callback function for get_config (used for profiling)
  - IncrementalUpdateGPPosteriorState
  - Fix of SkipPeriodicallyPredicate
  - default_next_candidates_algorithm

Since 0.1.0 seems released, this should go in now.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Innixma** [comment]
Clarification on this PR: v0.1 is not yet released. It is planned to be released sometime in the next 2-3 days.

Lets avoid review until v0.1 is fully released.

**Comment by szha** [comment]
Job PR-971-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-971/2/index.html

File: core/src/autogluon/core/scheduler/remote/remote.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ class Remote(Client):
     LOCK = mp.Lock()
     REMOTE_ID = mp.Value('i', 0)
     def __init__(self, remote_ip=None, port=None, local=False, ssh_username=None,
-            ssh_port=22, ssh_private_key=None, remote_python=None, timeout=30):
+            ssh_port=22, ssh_private_key=None, remote_python=None, timeout=60):
```
Comment (ValerioPerrone) [inline core/src/autogluon/core/scheduler/remote/remote.py:64]: What is the context for increasing the timeout?

File: core/src/autogluon/core/searcher/bayesopt/autogluon/gp_multifidelity_searcher.py [hunk 2]
```diff
@@ -1,36 +1,33 @@
...
+    DEFAULT_NUM_INITIAL_RANDOM_EVALUATIONS
+from ..utils.debug_log import DebugLogPrinter
 from ..utils.duplicate_detector import DuplicateDetectorIdentical
+from ..utils.simple_profiler import SimpleProfiler
 
 logger = logging.getLogger(__name__)
 
 
-class GPMultiFidelitySearcher(object):
+class GPMultiFidelitySearcher(GPSearcher):
```
Comment (ValerioPerrone) [inline core/src/autogluon/core/searcher/bayesopt/autogluon/gp_multifidelity_searcher.py:30]: Nice, looks much cleaner.

File: core/src/autogluon/core/searcher/bayesopt/autogluon/gp_multifidelity_searcher.py [hunk 3]
```diff
@@ -104,90 +102,39 @@ def __init__(
             this many candidates received label feedback
         :param initial_scoring: Scoring function to rank initial candidates.
             Default: thompson_indep (independent Thompson sampling)
-        :param profiler: If given, HPO computations are profiled
         :param first_is_default: If true, the first result of get_config is the
             default config of hp_ranges
         :param debug_log: DebugLogPrinter for debug logging (optional)
+        :param profiler: If given, HPO computations are profiled
```
Comment (ValerioPerrone) [inline core/src/autogluon/core/searcher/bayesopt/autogluon/gp_multifidelity_searcher.py:108]: cost_metric_name is missing from the :param list.

File: core/src/autogluon/core/searcher/bayesopt/autogluon/gp_fifo_searcher.py [hunk 4]
```diff
@@ -200,18 +165,116 @@ def update(self, config: Candidate, reward: float, **kwargs):
...
+            acquisition_class: Type[AcquisitionFunction],
+            init_state: TuningJobState = None,
+            local_minimizer_class: Type[LocalOptimizer] = DEFAULT_LOCAL_OPTIMIZER_CLASS,
+            skip_optimization: SkipOptimizationPredicate = None,
+            num_initial_candidates: int = DEFAULT_NUM_INITIAL_CANDIDATES,
+            num_initial_random_choices: int = DEFAULT_NUM_INITIAL_RANDOM_EVALUATIONS,
+            initial_scoring: Optional[str] = None,
+            first_is_default: bool = True,
+            debug_log: Optional[DebugLogPrinter] = None,
+            cost_metric_name: Optional[str] = None,
```
Comment (ValerioPerrone) [inline core/src/autogluon/core/searcher/bayesopt/autogluon/gp_fifo_searcher.py:232]: cost_metric_name is missing a :param description below.

**Review by ValerioPerrone — APPROVED** [review]
Looks good to me.

File: core/src/autogluon/core/scheduler/remote/remote.py (hunk 1 above)
Comment (mseeger) [reply core/src/autogluon/core/scheduler/remote/remote.py:64]: We are still seeing failures to connect, and some posts on Dask recommend to increase the timeout

File: core/src/autogluon/core/searcher/bayesopt/autogluon/gp_multifidelity_searcher.py (hunk 3 above)
Comment (mseeger) [reply core/src/autogluon/core/searcher/bayesopt/autogluon/gp_multifidelity_searcher.py:108]: OK

File: core/src/autogluon/core/searcher/bayesopt/autogluon/gp_fifo_searcher.py (hunk 4 above)
Comment (mseeger) [reply core/src/autogluon/core/searcher/bayesopt/autogluon/gp_fifo_searcher.py:232]: OK

**Comment by szha** [comment]
Job PR-971-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-971/3/index.html

**Comment by mseeger** [comment]
Can I merge this, or somebody else wants to have a look?

**Comment by Innixma** [comment]
@gradientsky Could you take a look?

**Comment by szha** [comment]
Job PR-971-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-971/4/index.html

File: core/src/autogluon/core/scheduler/hyperband.py [hunk 5]
```diff
@@ -673,26 +685,39 @@ def load_state_dict(self, state_dict):
...
+                level = 0 if reported_result is None \
+                    else reported_result[self._time_attr]
+                # It is possible to have tasks in _running_tasks which have
+                # reached self.max_t. These must not end up in the snapshot
+                if level < self.max_t:
+                    tasks[k] = {
+                        'config': v['config'],
+                        'time': v['time_stamp'],
+                        'level': level}
+        return tasks
```
Comment (gradientsky) [inline core/src/autogluon/core/scheduler/hyperband.py:705]: I'm not sure I understand the full flow, just wanted to get some clarity how the last block triggered after the job is done in the last bracket. Specifically, we're interested not losing the last values.

Comment (mseeger) [reply core/src/autogluon/core/scheduler/hyperband.py:705]: The results for an experiment are collected in training_history, as is done for all schedulers, and there all reported results are always collected.

Snapshots are a pretty specific feature only needed for specific searchers. These searchers need to know which tasks are currently still running and will report at least one further value. This is why the tasks which have reached max_t already, are filtered out here.

**Comment by mseeger** [comment]
Do we have any progress here?

**Comment by Innixma** [comment]
Let me try re-running CI, I was waiting for CI to pass but it looks like it may have been from a transient issue.

**Comment by szha** [comment]
Job PR-971-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-971/6/index.html

**Review by gradientsky — APPROVED** [review]
